### PR TITLE
Hdxdsys 865 build cesa pipeline

### PR DESCRIPTION
**Pull request by turnerm**
Here is a first implementation of the CESA pipeline. We're still waiting for feedback from them but hopefully the changes they request won't be very large.

In addition to the pipeline itself, I wanted to get your feedback on the overarching repo structure. Some things to note:

* I moved everything relevant from CONTRIBUTING.md to the readme, just because the readme was not very large in this case so I thought it saves having an extra file - perhaps that's something we can do for the scrapers. But I can put it back if we really want to be consistent through all our repos. 
* I wanted to be able to be able to throw in the requirements loosely and then generate a fixed requirements file. I did this using pyproject.toml but it adds a bit of complexity because you have to specify some things about the package, etc for it to all work.
* Related, I generated a separate requirements-test.txt since the test packages aren't needed for deployment, but then maybe we have too many files.

Curious to hear your preferences on the above, I'm not married to any of my decisions so happy to converge on what the team likes best.

### CODE REVIEW

File: cesa.py
Comment (b-j-mills) [inline cesa.py]: slugify defaults to lower case, so you could remove .lower()

File: .github/workflows/run-python-script.yaml
Comment (b-j-mills) [inline .github/workflows/run-python-script.yaml:20]: You might need to install GDAL separately like: https://github.com/OCHA-DAP/hdx-scraper-chirps/blob/68246b6ab3f1f52e2a23bdc71f80a24553bbe333/.github/workflows/run-python-script.yaml#L20
I also have had trouble with different python versions on GitHub Actions working with GDAL, so in my example I fixed python at 3.9.

File: .github/workflows/run-python-tests.yaml
Comment (b-j-mills) [inline .github/workflows/run-python-tests.yaml]: Similarly might need to install GDAL here

File: .config/hdx_dataset_static.yaml
Comment (b-j-mills) [inline .config/hdx_dataset_static.yaml]: You could go the opposite way and put more of the config into this, like notes, maintainer, org, update frequency, subnational, etc.  Then leave the project config really bare bones with only the url and tags.  You could then remove some lines from cesa.py too.

**Review by b-j-mills — APPROVED** [review]
Looks great!

File: run.py
Comment (mcarans) [inline run.py]: I would use os.path.join here rather than using /

File: .github/workflows/run-python-tests.yaml
Comment (mcarans) [inline .github/workflows/run-python-tests.yaml]: You can use the GH Action to install Hatch:
```
    - name: Install Hatch
      uses: pypa/hatch@install
```

File: .github/workflows/run-python-tests.yaml
Comment (mcarans) [inline .github/workflows/run-python-tests.yaml]: Moving to new Hatch features, you an do:
```
      run: |
        hatch test
    - name: Check styling
      if: always()
      run: |
        hatch fmt --check
```
(assuming you make the other changes I suggest in pyproject.toml)

File: pyproject.toml
Comment (mcarans) [inline pyproject.toml]: Using the new Hatch features, you can change to:
```
[tool.hatch.envs.hatch-test]
features = ["test"]

[[tool.hatch.envs.hatch-test.matrix]]
python = ["3.12"]

[tool.hatch.envs.hatch-test.scripts]
run = """
       pytest -c .config/pytest.ini --rootdir=. --junitxml=test-results.xml \
       --cov --cov-config=.config/coveragerc --no-cov-on-fail \
       --cov-report=lcov --cov-report=term-missing
       """

[tool.hatch.envs.hatch-static-analysis.scripts]
format-check = ["ruff format --config .config/ruff.toml --check --diff {args:.}",]
format-fix = ["ruff format --config .config/ruff.toml {args:.}",]
lint-check = ["ruff check --config .config/ruff.toml {args:.}",]
lint-fix = ["ruff check --config .config/ruff.toml --fix {args:.}",]
```

(That is provided you make the changes to the GH Actions workflow I suggested)

**Review by mcarans — APPROVED** [review]
I also moved to something like this setup in https://github.com/OCHA-DAP/hdx-scraper-hno so fully agree that we should use this kind of setup instead of what all other existing scrapers use.

If CONTRIBUTING.md is something constant between scrapers then could your cookiecutter generate it? If so, that might be a reason to keep it separate although I guess a README.md could be generated which contains this text. Whatever is easier for the cookiecutter is fine with me.

I think using minimal requirements for running and a larger test requirements is probably a good habit to have. hapi-pipelines generates prod-requirements.txt using the versions from the larger requirements.txt in the publish step: [`pip-compile pyproject.toml --resolver=backtracking --upgrade -q -c requirements.txt -o prod-requirements.txt`](https://github.com/OCHA-DAP/hapi-pipelines/blob/main/.github/workflows/publish.yaml#L41). That negates the need to check in two lots of requirements files (and ensures that what is published is using the same versions as what has been tested).

I envisaged that the `.config` folder contains build stuff that is nearly all common between scrapers. I think it would be better to keep the files like `project_configuration.yaml` and `hdx_dataset_static.yaml` separate as these contain scraper specific stuff that is required for running the Python script. 

In hdx-scraper-hno I have used a `src` folder and the scraper is run as a module: `python -m hdx.scraper.hno`. This works better for more complex scrapers that need multiple files. Should we make that the general form to accommodate all scrapers or keep to the simpler form that suits many scrapers? Note that when running as a module, I found it easier to locate the config yaml files in the running script by putting them in the same folder as the Python files, but am open to using a separate `config` or `configuration` folder if that is clearer (and can be made to work reliably).

File: .github/workflows/run-python-script.yaml
Comment (turnerm) [reply .github/workflows/run-python-script.yaml:20]: Yes good point I was thinking that, I've done it in the past but my way was a [bit more comples](https://github.com/OCHA-DAP/IBF_TYPHOON_DATA_PIPELINE/blob/main/.github/workflows/test-typhoon-model.yml#L40) so I'll give yours a shot if needed. I will wait to see if the pipeline complains, so far the tests seem to work at least.

File: .config/hdx_dataset_static.yaml
Comment (turnerm) [reply .config/hdx_dataset_static.yaml]: Ah I didn't realize that those could all go in the static config, I will move them, thanks!

File: run.py
Comment (turnerm) [reply run.py]: Oops you're right thanks!

**Comment by turnerm** [comment]
@mcarans 

> If CONTRIBUTING.md is something constant between scrapers then could your cookiecutter generate it? If so, that might be a reason to keep it separate although I guess a README.md could be generated which contains this text. Whatever is easier for the cookiecutter is fine with me.

My thinking was the CONTRIBUTING.md makes sense in packages like hdx-python-api where we expect there to be  external users of the README and we may not want it bogged down with all the technical details. The CONTRIBUTING.md is a lot longer in that case as well. 

One downside of keeping them separate though is that if someone does want to contribute, they may not realize that the information they need is there - I think I had to point Ian to it in one of our repos, for example. Perhaps it should be at least linked to from the README. 

But I digress -- for the pipelines which I assume are always internal, to me it makes sense to combine them for efficiency. The README can be pretty much boilerplate except for more complex pipelines. 

However happy to defer to other preferences, I just wanted to explain my reasoning. 

> I think using minimal requirements for running and a larger test requirements is probably a good habit to have. hapi-pipelines generates prod-requirements.txt using the versions from the larger requirements.txt in the publish step: [pip-compile pyproject.toml --resolver=backtracking --upgrade -q -c requirements.txt -o prod-requirements.txt](https://github.com/OCHA-DAP/hapi-pipelines/blob/main/.github/workflows/publish.yaml#L41). That negates the need to check in two lots of requirements files (and ensures that what is published is using the same versions as what has been tested).

Sounds good, I will keep the two requirements files and generate the test one from requirements.txt. Actually I think that's better because otherwise you can get into weird situations where the two fixed requirements files don't quite match - happened to me with AnticiPy. 

> I envisaged that the .config folder contains build stuff that is nearly all common between scrapers. I think it would be better to keep the files like project_configuration.yaml and hdx_dataset_static.yaml separate as these contain scraper specific stuff that is required for running the Python script.

Makes sense, I will move them. 

> In hdx-scraper-hno I have used a src folder and the scraper is run as a module: python -m hdx.scraper.hno. This works better for more complex scrapers that need multiple files. Should we make that the general form to accommodate all scrapers or keep to the simpler form that suits many scrapers? Note that when running as a module, I found it easier to locate the config yaml files in the running script by putting them in the same folder as the Python files, but am open to using a separate config or configuration folder if that is clearer (and can be made to work reliably).

No I think I like being consistent, even if it's slight overkill for the simple pipelines it's just nicer and avoids annoying problems like getting hatch to work without a real build etc. I will change this scraper to match the above format. 

By the way I found a tinty gitignore bug in the HNO pipeline: https://github.com/OCHA-DAP/hdx-scraper-hno/blob/main/.gitignore#L164

**Pull request merged**
